src/com/amazon/ionschema/internal/constraint/Regex.kt

```diff
@@ -21,55 +21,208 @@ import com.amazon.ion.IonValue
+                }
+                else -> sb.append(ch)           // otherwise, accept the character
+            }
+
+            parseQuantifier(si, sb)   // parse a quantifier, if present
+
+            ch = si.next()
+        } while (ch != null)
+
+        return Pattern.compile(sb.toString(), flags)
+    }
+
+    private fun parseCharacterClass(si: StringIterator, sb: StringBuilder) {
+        var ch = si.next()
+        sb.append(ch)
+
+        if (ch == '^') {
+            ch = si.next()
+            sb.append(ch)
+        }
+
+        ch = si.next()
+        if (ch == '\\') {      // block use of Pattern's "[\p..." constructs
+            error(si, "invalid character '$ch' in character class")
+        }
+        sb.append(ch)
+
+        if (ch == '-') {       // it's a character range
+            ch = si.next()
+            sb.append(ch)
+
+            ch = si.next()
+            when (ch) {
+                ']' -> sb.append(ch)
+                null -> error(si, "unexpected end of string")
+                else -> error(si, "unexpected character '$ch'")
+            }
+
+        } else {               // it's a character class
+            var complete = false
+            ch = si.next()
+            while (ch != null && !complete) {
+                sb.append(ch)
+                if (ch == ']') {
+                    complete = true
+                } else {
+                    ch = si.next()
+                }
+            }
+
+            if (!complete) {
+                error(si, "character class missing ']'")
+            }
+        }
+    }
+
+    private fun parseQuantifier(si: StringIterator, sb: StringBuilder) {
+        val initialLength = sb.length
+        var ch = si.peek()
+        when (ch) {
+            '?', '*', '+' -> {
+                ch = si.next()
+                sb.append(ch)
+            }
+            '{' -> {
+                ch = si.next()
+                sb.append(ch)
+                var complete = false
+                do {
+                    ch = si.next()
+                    when (ch) {
+                        '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', ',' -> sb.append(ch)
+                        '}' -> {
+                            sb.append(ch)
+                            complete = true
+                        }
+                        null -> {}
+                        else -> error(si, "invalid character '$ch'")
+                    }
+                } while (ch != null && !complete)
+
+                if (!complete) {
+                    error(si, "range quantifier missing '}'")
+                }
+            }
+        }
+
+        if (sb.length > initialLength && ch != null) {
+            ch = si.peek()
+            when (ch) {
+                '?' -> error(si, "invalid character '$ch'")
+                '+' -> error(si, "invalid character '$ch'")
+            }
+        }
+    }
+
+    private fun error(si: StringIterator, message: String): Unit =
```

Comment:
nit: I find it useful to add return types on method/function signatures. Once we make a decision then we should be consistent including methods that have no return value, i.e., `Unit`

Reply:
I've been declaring return types for public methods for clarity, but omitting them for anything internal.  In this instance, the compiler requires the return type.

Reply:
Yes the compiler does require it. Which is why it makes me consider the option of adding it regardless. Internal developers (us) should get as much help as external developers. 

Not a deal breaker since we are already non-consistent.